tests/DCPS/Presentation/main.cpp

```diff
@@ -8,350 +10,584 @@
+          ACE_TEXT("delete subscriber failed: %C\n"),
+          retcode_to_dds_string(rc).c_str()));
       }
-      ACE_OS::sleep(5); // wait for samples to arrive
-
-      // Verify that no samples are yet available
-      {
-        FooSeq foo;
-        DDS::SampleInfoSeq info;
-
-        DDS::ReturnCode_t error;
-        if ((error = reader_i->take(foo,
-                                    info,
-                                    DDS::LENGTH_UNLIMITED,
-                                    DDS::ANY_SAMPLE_STATE,
-                                    DDS::ANY_VIEW_STATE,
-                                    DDS::ANY_INSTANCE_STATE)) != DDS::RETCODE_NO_DATA)
-        {
-          ACE_ERROR_RETURN((LM_ERROR,
-                            ACE_TEXT("%N:%l main()")
-                            ACE_TEXT(" ERROR: Unexpected samples read!\n")), -1);
-        }
+    }
+
+    if (!CORBA::is_nil(publisher_)) {
+      rc = participant_->delete_publisher(publisher_);
+      if (rc != DDS::RETCODE_OK) {
+        ACE_ERROR((LM_ERROR, ACE_TEXT("%N:%l ~Test() ERROR: ")
+          ACE_TEXT("delete publisher failed: %C\n"),
+          retcode_to_dds_string(rc).c_str()));
       }
+    }
+  }
+};
 
-      publisher->end_coherent_changes();
-      ACE_OS::sleep(5); // wait for samples to arrive
-
-      //Verify coherent samples are now available
-      {
-        FooSeq foo;
-        DDS::SampleInfoSeq info;
-
-        DDS::ReturnCode_t error;
-        if ((error = reader_i->take(foo,
-                                    info,
-                                    DDS::LENGTH_UNLIMITED,
-                                    DDS::ANY_SAMPLE_STATE,
-                                    DDS::ANY_VIEW_STATE,
-                                    DDS::ANY_INSTANCE_STATE)) != DDS::RETCODE_OK)
-        {
-          ACE_ERROR_RETURN((LM_ERROR,
-                            ACE_TEXT("%N:%l main()")
-                            ACE_TEXT(" ERROR: take failed %d\n"), error), -1);
-        }
+bool
+coherent_test(DDS::DomainParticipant_var& participant, DDS::Topic_var& topic)
+{
+  DDS::ReturnCode_t rc;
+
+  ACE_DEBUG((LM_DEBUG, ACE_TEXT("Testing coherent_access...\n")));
+
+  DDS::Publisher_var publisher;
+  DDS::Subscriber_var subscriber;
+  DDS::DataReader_var reader1;
+  DDS::DataReader_var reader2;
+  DDS::DataWriter_var writer;
+  Test t(participant, publisher, subscriber, reader1, reader2, writer);
+  if (t.setup_test(topic, true)) {
+    ACE_ERROR_RETURN((LM_ERROR, ACE_TEXT("%N:%l coherent_test() ERROR: ")
+                      ACE_TEXT("setup_test failed!\n")), true);
+  }
 
-        if (foo.length() != SAMPLES_PER_TEST)
-        {
-          ACE_ERROR_RETURN((LM_ERROR,
-                            ACE_TEXT("%N:%l main()")
-                            ACE_TEXT(" ERROR: Unexpected number of samples taken: %d\n"), foo.length()), -1);
-        }
-      }
+  FooDataReader_var reader1_i = FooDataReader::_narrow(reader1);
+  if (CORBA::is_nil(reader1_i)) {
```

Comment:
```suggestion
  if (!reader1_i) {
```
As long as you're overhauling this test I'd change all the `is_nil`s to the simpler form.

Reply:
I was wondering if that worked, but I couldn't remember so I just went with what the test already had.